x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -0,0 +1,211 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.profile;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.get.GetAction;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.search.SearchAction;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.security.action.profile.Profile;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.SECURITY_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_PROFILE_ALIAS;
+
+public class ProfileService {
+    private static final Logger logger = LogManager.getLogger(ProfileService.class);
+    private static final String DOC_ID_PREFIX = "profile_";
+
+    private final Settings settings;
+    private final Clock clock;
+    private final Client client;
+    private final SecurityIndexManager profileIndex;
+    private final ThreadPool threadPool;
+
+    public ProfileService(Settings settings, Clock clock, Client client, SecurityIndexManager profileIndex, ThreadPool threadPool) {
+        this.settings = settings;
+        this.clock = clock;
+        this.client = client;
+        this.profileIndex = profileIndex;
+        this.threadPool = threadPool;
+    }
+
+    public void getProfile(String uid, @Nullable Set<String> dataKeys, ActionListener<Profile> listener) {
+        getVersionedDocument(uid, ActionListener.wrap(versionedDocument -> {
+            // TODO: replace null with actual domain lookup
+            listener.onResponse(versionedDocument.toProfile(null, dataKeys));
+        }, listener::onFailure));
+    }
+
+    private void getVersionedDocument(String uid, ActionListener<VersionedDocument> listener) {
+        if (maybeHandleIndexStatusIssue(listener)) {
```

Comment:
I would think there's value in maintaining the old pattern of freezing the index at the method start, and using the freezed reference in all the following calls.
Plus it avoids coding methods like this which return a value and also signal to the listener.

Reply:
The method does freeze the index and does the checking. But I forgot to return the frozen index while refactoring. I don't like repeating this pattern in multiple methods and that's the motivation of extracting it. Because of the callback style coding, this kinda extraction always leads to the non-ideal situation where a function returns a value and also signal the listener. It's a tradeoff between code reuse and stay true to callback.

I'd really like to avoid the repetition of the pattern. So here is another attempt. I renamed the method to `tryFreezeAndCheckIndex` which returns an `Optional<SecurityIndexManager>`. So the code becomes:

```
tryFreezeAndCheckIndex(listener).ifPresent(frozenProfileIndex -> {
    // do stuff
});
```
I think it's a better balance between the tradeoffs. What do you think?

Reply:
Your call, I don't have much of an opinion.